Source/SwiftLintBuiltInRules/Rules/Lint/MarkRule.swift

```diff
@@ -9,239 +10,143 @@ struct MarkRule: CorrectableRule {
+        private static let twoOrThreeSlashes = "///?"
+        private static let colonOrEmpty = ":?"
+        private static let hyphenOrEmpty = "-? *"
 
-        result.append(contentsOf: correct(file: file,
-                                          pattern: threeSlashesInsteadOfTwo,
-                                          replaceString: "// MARK:"))
+        private static func nonCapturingGroup(_ pattern: String) -> String {
+            "(?:\(pattern))"
+        }
 
-        return result.unique
-    }
+        private static func capturingGroup(_ pattern: String) -> String {
+            "(\(pattern))"
+        }
 
-    private func correct(file: SwiftLintFile,
-                         pattern: String,
-                         replaceString: String,
-                         keepLastChar: Bool = false) -> [Correction] {
-        let violations = violationRanges(in: file, matching: pattern)
-        let matches = file.ruleEnabled(violatingRanges: violations, for: self)
-        if matches.isEmpty { return [] }
-
-        var nsstring = file.contents.bridge()
-        let description = Self.description
-        var corrections = [Correction]()
-        for var range in matches.reversed() {
-            if keepLastChar {
-                range.length -= 1
+        private static func basePattern(_ pattern: String) -> String {
+            nonCapturingGroup("\(twoOrThreeSlashes)\(anySpace)\(pattern)\(anySpace)\(colonOrEmpty)\(anySpace)")
+        }
+
+        private static func replace(_ target: String, range nsrange: NSRange, to replaceString: String) -> String {
+            guard nsrange.length > 0, let range = Range(nsrange, in: target) else {
+                return target
             }
-            let location = Location(file: file, characterOffset: range.location)
-            nsstring = nsstring.replacingCharacters(in: range, with: replaceString).bridge()
-            corrections.append(Correction(ruleDescription: description, location: location))
+            return target.replacingCharacters(in: range, with: replaceString)
         }
-        file.write(nsstring.bridge())
-        return corrections
     }
 
-    private func violationRanges(in file: SwiftLintFile, matching pattern: String) -> [NSRange] {
-        return file.rangesAndTokens(matching: pattern).filter { matchRange, syntaxTokens in
-            guard
-                let syntaxToken = syntaxTokens.first,
-                let syntaxKind = syntaxToken.kind,
-                SyntaxKind.commentKinds.contains(syntaxKind),
-                case let tokenLocation = Location(file: file, byteOffset: syntaxToken.offset),
-                case let matchLocation = Location(file: file, characterOffset: matchRange.location),
-                // Skip MARKs that are part of a multiline comment
-                tokenLocation.line == matchLocation.line
-            else {
-                return false
+    func violationResults(withCorrect: Bool = false) -> [ViolationResult] {
```

Comment:
It seems like you don't need the parameter `withCorrect`, because all corrections are evaluated lazily anyway. You need to move the `correct()` call into the correction closure though.